### ISSUE TITLE
Enhance request/response abilities of net_response input plugin

### ISSUE THREAD

**Issue by jjbuchan** [influxdata/telegraf#6873]
## Feature Request

An input plugin that can send multiple strings to a tcp connection while also reading the response of each action.

### Proposal:

Create a new `net_response_v2` input plugin to eventually replace [net_response](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/net_response).  Instead of using single values, it will accept a list of `send` and `expect` strings, and additionally add an `expect_banner` option and a `send_on_failure` option.

The `expect_banner` (if set) will be checked after the initial connection is made.
The strings in the `expect` list will be checked immediately after any `send` string has been executed.
If one of those `expect` strings is not found when expected the `send_on_failure` string will be sent if specified, otherwise the connection will be closed.

#### Questions

This seems like a good tactic so that new functionality can be introduced in a minor version change without impacting any existing uses and without complicating a single plugin to handle both conigurations, but I am not sure whether this adheres to any prior standard for plugin enhancements?

I would assume the v2 version is initially included in addition to the existing `net_response` plugin in a minor version update, but would entirely replace it on the next major version change?

### Current behavior:

The `net_response` input plugin currently only allows for single `send` and `expect` to be set.  If a response is received immediately after the connection is opened the `expect` string will be tested against that.  If a `send` string is then executed there is no way to validate the output of that command.

### Desired behavior:

The ability to execute any number of commands against a tcp connection while also validating their response.

#### SMTP
One example is to monitor the smtp protocol, where you want to verify the initial 220 code as well as the 250 and 221.

```
$ telnet mysmtptest.com 25

Trying mysmtptest.com...
Connected to mysmtptest.com.
Escape character is '^]'.
220 mysmtptest.com ESMTP Postfix

> ehlo me.com
250-mysmtptest.com
250-PIPELINING
250-SIZE 10240000
250-VRFY
250-ETRN
250-AUTH PLAIN LOGIN
250-ENHANCEDSTATUSCODES
250-8BITMIME
250 DSN

> quit
221 2.0.0 Bye
```

The plugin may be configured with the following parameters.

```
expect_banner = "220"
send = ["ehlo me.com", "quit"]
expect = ["250", "221"]
send_on_failure = "quit"
```

In this case if the `250` response is not seen after sending the `ehlo` command a `quit` will still be sent regardless in an attempt to cleanly close the connection.  For this example it does not make much difference, but there may be cases where continuing to execute the `send` commands after a failure is seen is problematic.

#### FTP
Another is ensuring ftp connections closed out as expected, by verifying the initial 220 and the final 221.

```
$ telnet myftptest.com 21
Trying myftptest.com...
Connected to myftptest.com.
Escape character is '^]'.
220 Hello.

> quit
221 Goodbye.
```

The plugin may be configured with the following parameters.

```
expect_banner = "220"
send = ["quit"]
expect = ["221"]
```

#### Custom Services
and then of course there is any custom service that you may have running to accepts input and respond appropriately.  These may or may not have an initial banner.  In this example it connects with no banner, issues an `info` command, and expects to see a `connected_clients` string.

```
$ telnet mycustomerservice.com 3125
Trying mycustomerservice.com...
Connected to mycustomerservice.com.
Escape character is '^]'.

> info
connected_clients:18
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:0
```

The plugin may be configured with the following parameters.

```
send = ["info"]
expect = ["connected_clients"]
```

**Comment by jjbuchan** [comment]
A secondary request would be to add SSL support to this plugin as well.

**Comment by jjbuchan** [comment]
@danielnelson could you provide some feedback for this proposal?  I should be able to commit to the work but would like some buy-in before starting work on a PR.

Thanks

**Comment by danielnelson** [comment]
Are you looking to collect the same metrics as before (response_time and match/nomatch)?

I believe all of these examples could be done with the current `net_response` plugin, enough to tell if the service is up.  I'm unsure if we should try to go beyond checking if a service is up with this plugin,  do we really need to automate an entire SMTP session?  For more complicated protocols, I'm still of the opinion that it makes the most sense to have specific plugins, or to fallback to the exec input.

Adding TLS seems like a good non-controversial feature.

**Comment by jjbuchan** [comment]
Yep, looking for the same metrics.

Of the three examples I gave I can definitely see the full SMTP session being its own plugin.  I can create a new issue for a specific smtp proposal if you're welcoming of that idea?

We could also get by with using the current plugin for the FTP example as it is most important that we send the `quit` command and less so that we verifying the `221` code.

For the final case, after a little more investigation I can see where we were having problems.  The plugin currently [only reads the first line of the response](https://github.com/influxdata/telegraf/blob/9d214ae291ba4e7e7df66c91fb50460a01a9e4e5/plugins/inputs/net_response/net_response.go#L111).  In some cases we were not appending `\n\n` to the `send` to trigger it to submit and in others we were expecting a string that wasn't on the first line.  To try and alleviate the latter, instead of performing a `readLine` could we instead read up to 1024 bytes, convert that to a string, and then compare the `expect` against it?  This would help for cases where you receive an initial response upon connection but actually want to validate the response of the `send`.

**Comment by sjwang90** [comment]
Check out #8077 and if it touches what you're looking for.

**Comment by Hipska** [comment]
Hi, this sounds a reasonable request and PRs are welcome.

IMHO I would not create a separate plugin, the new options could be added backwards compatible while deprecating the old `send` and `expect` strings.
